### PR TITLE
Render the Contao login page even if the current route is not a Contao page

### PR DESCRIPTION
**Pull request by aschempp**
Fixed point 1 of https://github.com/contao/contao/pull/6805 for Contao 4.13

This must not be merged upstream!

### CODE REVIEW

**Comment by fritzmg** [comment]
I am not quite sure what to test here. Do you mean a custom controller that has a route starting with `%contao.backend.route_prefix%` but has no `backend` `_scope`?

**Comment by fritzmg** [comment]
I tried with 

```php
namespace App\Controller;

use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Annotation\Route;

#[Route(path: '%contao.backend.route_prefix%/test')]
class TestController
{
    public function __invoke(): Response
    {
        return new Response('Hello World!');
    }
}
```

But this controller is always rendered, regardless of whether you have a valid back end login or not (as it should, I think?).

**Comment by aschempp** [comment]
This is not about the back end. I have a custom controller on a custom route, that has `scope = frontend`. The controller checks permissions using Symfony, which works fine because the front end firewall is active. But because it does not have a `pageModel` in the request attributes, the Contao 401 login page is not rendered if access is denied.

**Comment by fritzmg** [comment]
Ah, I see. I have the same issue within `fritzmg/contao-file-access` - I hacked it in by [executing](https://github.com/fritzmg/contao-file-access/blob/67051626e057c2c00390dedc65769cea99456ca1/src/Controller/AbstractFilesController.php#L96-L105)

```php
$root = $this->findFirstPublishedRootByHostAndLanguage($request->getHost(), $request->getLocale());

if (null !== $root) {
    $root->loadDetails();
    $request->attributes->set('pageModel', $root);
    $GLOBALS['objPage'] = $root;
}
```

This means after this PR this should not be necessary anymore, correct?

**Review by fritzmg — CHANGES_REQUESTED** [review]
I tested it - but it does not work if you set `contao.legacy_routing: false`. When Legacy Routing is disabled, `Frontend::getRootPageFromUrl()` will simply check for the `pageModel` attribute in the current request, which is of course missing.

https://github.com/contao/contao/blob/d26e0cec7adaf99a6f10c55db5449078ead30552/core-bundle/src/Resources/contao/classes/Frontend.php#L302-L321

**Comment by aschempp** [comment]
Thanks @fritzmg. I have updated the PR to use other means to find the root page. The used method is deprecated, but since this is not going to be merged upstream I think that should be fine.

File: core-bundle/src/EventListener/PrettyErrorScreenListener.php [hunk 1]
```diff
@@ -250,4 +247,25 @@ private function getStatusCodeForException(\Throwable $exception): int
...
     }
+
+    private function findErrorPage(int $type, Request $request): PageModel|null
+    {
+        $pageModel = $request->attributes->get('pageModel');
+
+        if (!$pageModel instanceof PageModel) {
+            $rootRequest = Request::create('http://'.$request->getHost());
+            $rootRequest->headers->set('Accept-Language', $rootRequest->headers->get('Accept-Language'));
+            $parameters = $this->requestMatcher->matchRequest($request);
```
Comment (fritzmg) [inline core-bundle/src/EventListener/PrettyErrorScreenListener.php]: Right now it works neither with nor without legacy routing. I think this needs to be changed?

```suggestion
            $parameters = $this->requestMatcher->matchRequest($rootRequest);
```

**Comment by fritzmg** [comment]
> The used method is deprecated

Btw. which method do you mean? As far as I can see you did not use any deprecated methods 🤔

**Review by fritzmg — DISMISSED** [review]

**Comment by leofeyer** [comment]
Wouldn‘t it be easier to backport the `PageFinder` service? This would also prevent conflicts when merging upstream.

**Comment by aschempp** [comment]
> Wouldn‘t it be easier to backport the `PageFinder` service? This would also prevent conflicts when merging upstream.

I don‘t think its a good idea to backport this so late in the patch process…

**Comment by leofeyer** [comment]
Why not? It wouldn‘t break anything and it would ensure upstream mergeability, which is worth much more than following the "no new features in a bugfix release" rule. We could even make the service `@internal` in Contao 4.13, so that nobody uses it, if that is what you‘re worried about.

**Comment by fritzmg** [comment]
To be fair, it could also reduce duplicate code in Contao 4.13.

**Comment by aschempp** [comment]
The `PageFinder` is a feature, it allows a lot more than this bug fix requires. Not sure how we could fix the upstream mergeability issue, but that's a one-time thing, right? I would
1. merge everything before this PR upstream
2. merge this PR
3. merge upstream again, and remove all changes

That should do it?

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @aschempp.

**Pull request merged**